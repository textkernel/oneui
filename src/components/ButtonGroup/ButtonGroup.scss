@import '../../themes/oneui/oneui.scss';

.ButtonGroup {
    display: inline-flex;

    &__button {
        border-radius: 0;

        &--first {
            border-top-left-radius: var(--border-radius);
            border-bottom-left-radius: var(--border-radius);
        }

        &--last {
            border-top-right-radius: var(--border-radius);
            border-bottom-right-radius: var(--border-radius);
        }

        ~ .ButtonGroup__button {
            margin-left: -1px;
        }


        &:not(.ButtonGroup__button--first):not(:focus) {
            border-left-color: var(--color-neutral-25);
        }

        &:not(.ButtonGroup__button--last):not(:focus) {
            border-right-color: var(--color-neutral-25);
        }

    }

    &__button--isPrimary {
        &:not(.ButtonGroup__button--first):not(:focus) {
            border-left-color: var(--color-primary-25);
        }

        &:not(.ButtonGroup__button--last):not(:focus) {
            border-right-color: var(--color-primary-25);
        }
    }

    &--isBlock {
        display: flex;
        width: 100%;
    }

    &__button--isBlock {
        flex-grow: 1;
    }
}
