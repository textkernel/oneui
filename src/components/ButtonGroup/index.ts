export { ButtonGroup, Props as ButtonGroupProps } from './ButtonGroup';
