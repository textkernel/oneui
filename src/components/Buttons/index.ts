export { Button, ButtonProps } from './Button';
export { FileButton, FileButtonProps } from './FileButton';
export { SearchButton } from './SearchButton';
export { StepperButton, StepperButtonProps } from './StepperButton';
