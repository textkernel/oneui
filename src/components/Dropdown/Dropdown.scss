@import '../../themes/oneui/oneui';

.Dropdown {
    &__list {
        background-color: var(--color-background);
        border-radius: var(--border-radius);
        border: var(--color-light) 1px solid;
        box-shadow: 0 5px 10px 0 rgba(0, 0, 0, 0.2);
        z-index: 400;

        .ListItem:hover:not(.ListItem--disabled) {
            background-color: var(--color-light-25);
        }
    }
}
