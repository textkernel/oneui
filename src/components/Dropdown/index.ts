export { Dropdown, Props as DropdownProps } from './Dropdown';
