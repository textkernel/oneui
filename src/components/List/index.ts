export { List, Props as ListProps } from './List';
export { ListItem, ListItemProps } from './ListItem';
export { ListActions, ListActionsProps } from './ListActions';
export { ListOptimizer, ListOptimizerProps } from './ListOptimizer';
