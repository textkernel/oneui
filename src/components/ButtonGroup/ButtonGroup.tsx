import * as React from 'react';
import { bem } from '../../utils';
import styles from './ButtonGroup.scss';
import { Context, Size } from '../../constants';
import { ButtonProps } from '../Buttons';

export interface Props {
    /** The buttons in this group */
    children: React.ReactElement<ButtonProps> | (React.ReactElement<ButtonProps> | EmptyElement)[];
    /** The context for all buttons in this group (e.g. brand, primary, bad, good etc.) */
    context?: Context | 'link';
    /** Whether or not to show block-level button group (full width) */
    isBlock?: boolean;
    /** The size of the buttons in the button group */
    size?: Size;
}

const { block, elem } = bem('ButtonGroup', styles);

export const ButtonGroup = ({
    children,
    context = 'neutral',
    size = 'normal',
    isBlock = false,
    ...rest
}: Props) => {
    const totalNumberOfButtons = React.Children.count(children);

    if (totalNumberOfButtons === 1) {
        return (
            <div {...rest} {...block({ isBlock })} role="group">
                {React.isValidElement(children)
                    ? React.cloneElement(children, {
                          ...children.props,
                          context,
                          size,
                          isBlock,
                      })
                    : children}
            </div>
        );
    }

    return (
        <div {...rest} {...block({ isBlock })} role="group">
            {React.Children.map(children, (button, i) => {
                if (!React.isValidElement(button)) {
                    return button;
                }

                return React.cloneElement(button, {
                    ...button.props,
                    context,
                    size,
                    ...elem('button', {
                        isBlock,
                        context,
                        first: i === 0,
                        last: i + 1 === totalNumberOfButtons,
                    }),
                });
            })}
        </div>
    );
};

ButtonGroup.displayName = 'ButtonGroup';
