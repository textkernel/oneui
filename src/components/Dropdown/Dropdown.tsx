import React, { useState } from 'react';
import { useSelect } from 'downshift';
import { usePopper } from 'react-popper';
import { EmptyElement } from '../../customTypes/types';
import { bem } from '../../utils/bem/bem';
import { mergeRefs } from '../../utils/mergeRefs';
import { PopupPlacement } from '../../constants';
import { ButtonProps } from '../Buttons';
import { ListItemProps } from '../List/ListItem';
import { List } from '../List';
import styles from './Dropdown.scss';

const { elem } = bem('Dropdown', styles);

export interface Props<V> extends Omit<React.HTMLAttributes<HTMLDivElement>, 'onChange'> {
    /**
     *  Popup content to be placed inside List.
     *  Navigation available only if element has not empty `value` attribute
     *  and empty/false `disabled` attribute.
     *  Use {@link ListItem} component.
     */
    children:
        | React.ReactElement<ListItemProps>
        | (React.ReactElement<ListItemProps> | EmptyElement)[]
        | (
              | React.ReactElement<ListItemProps>
              | React.ReactElement<ListItemProps>[]
              | EmptyElement
          )[];
    /** {@link Button} element, controlled by current component */
    button: React.FunctionComponentElement<ButtonProps>;
    /**
     * Callback called on selecting one of the passed as children items.
     * Value parameter it is a `value` attribute of children item ({@link ListItemProps.value}).
     */
    onChange: (value: V) => void;
    /**
     * Callback called on toggle button click with current/previous open state
     */
    onToggleClick?: (prevOpenState: boolean) => void;
    /**
     * Callback called on focus of the dropdown menu
     */
    onMenuFocus?: () => void;
    /**
     * Callback called on blur of the dropdown menu
     */
    onMenuBlur?: () => void;
    /**
     * Additional select props that enrich and use downshift API
     */
    additionalSelectProps?: {};
    /**
     * ClassName that is assigned to <ul> element of the dropdown
     */
    listClassName?: string;
    /** Popup placement relative to button */
    placement?: PopupPlacement;
}

/**
 * ## Usage information
 * This component is intended for displaying available actions by button click.
 * It allows you to navigate with keyboard across non-disabled actions.
 * If you don't need navigation - use PopupBase component.

 * Navigation available only through children which have not empty `value` attribute
 * and empty/false `disabled` attribute (all other children items will be skipped during navigation).
 * Use `ListItem` component as child item.
 */

export function Dropdown<V>({
    button,
    children,
    onChange,
    onToggleClick = undefined,
    onMenuBlur = undefined,
    onMenuFocus = undefined,
    placement = 'bottom-end',
    additionalSelectProps = {},
    listClassName = '',
    ...rest
}: Props<V>) {
    const [referenceElement, setReferenceElement] = useState(null);
    const [popperElement, setPopperElement] = useState(null);

    const state = usePopper(referenceElement, popperElement, {
        placement,
        modifiers: [
            {
                name: 'offset',
                options: { offset: [0, 3] },
            },
        ],
    });

    const childrenArray = React.Children.toArray(children);
    const valuesAvailableForHighlight: V[] = [];
    childrenArray.forEach((child) => {
        if (
            React.isValidElement(child) &&
            child.props.value !== undefined &&
            child.props.value !== null &&
            !child.props.disabled
        ) {
            valuesAvailableForHighlight.push(child.props.value);
        }
    });

    const { isOpen, getItemProps, getMenuProps, getToggleButtonProps, highlightedIndex, reset } =
        useSelect<V>({
            items: valuesAvailableForHighlight,
            onSelectedItemChange: ({ selectedItem }) => {
                if (selectedItem) {
                    onChange(selectedItem);
                    reset();
                }
            },
            ...additionalSelectProps,
        });

    const menuProps = getMenuProps();
    const toggleButtonProps = getToggleButtonProps({
        onClick: () => {
            onToggleClick?.(isOpen);
        },
        onBlur: onMenuBlur,
        onFocus: onMenuFocus,
    });
    const openPopperProps = isOpen && {
        style: state.styles.popper,
        ...state.attributes.popper,
        ...elem('list', {
            elemClassName: listClassName,
        }),
    };
    return (
        <>
            {React.cloneElement<any>(button, {
                ...rest,
                ...toggleButtonProps,
                ref: mergeRefs([setReferenceElement, toggleButtonProps.ref, button.ref]),
            })}
            <List
                {...menuProps}
                {...openPopperProps}
                ref={mergeRefs([isOpen ? setPopperElement : undefined, menuProps.ref])}
                isControlledNavigation
            >
                {isOpen &&
                    childrenArray.map((child) => {
                        if (
                            React.isValidElement(child) &&
                            valuesAvailableForHighlight.includes(child.props.value)
                        ) {
                            const currentValueIndex = valuesAvailableForHighlight.findIndex(
                                (val) => val === child.props.value
                            );

                            return React.cloneElement<any>(child, {
                                ...getItemProps({
                                    index: currentValueIndex,
                                    item: child.props.value,
                                }),
                                isHighlighted: highlightedIndex === currentValueIndex,
                            });
                        }
                        return child;
                    })}
            </List>
        </>
    );
}

Dropdown.displayName = 'Dropdown';
