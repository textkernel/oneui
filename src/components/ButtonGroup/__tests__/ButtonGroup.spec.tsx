import React from 'react';
import '@testing-library/jest-dom';
import { render, screen } from '@testing-library/react';
import { ButtonGroup } from '../ButtonGroup';
import { Button } from '../../Buttons';
import { Dropdown } from '../../Dropdown';
import { ListItem } from '../../List/ListItem';

describe('<ButtonGroup> that renders a button', () => {
    it('should render default button correctly', () => {
        const view = render(
            <ButtonGroup>
                <Button key="a">A button</Button>
                <Button key="b">Another button</Button>
            </ButtonGroup>
        );

        expect(view.container).toMatchSnapshot();
        const button = screen.getAllByRole('button');
        expect(button[0]).toHaveClass('ButtonGroup__button--first');
        expect(button[0]).not.toHaveClass('ButtonGroup__button--last');
        expect(button[1]).not.toHaveClass('ButtonGroup__button--first');
        expect(button[1]).toHaveClass('ButtonGroup__button--last');
    });

    it('should add classes when props are changed', () => {
        const view = render(
            <ButtonGroup size="large" isBlock>
                <Button key="a">A button</Button>
                <Button key="b">Another button</Button>
            </ButtonGroup>
        );

        expect(view.container).toMatchSnapshot();
        const button = screen.getAllByRole('button');
        expect(button[0]).toHaveClass('Button--size_large');
        expect(button[0]).toHaveClass('ButtonGroup__button--isBlock');
        expect(button[1]).toHaveClass('Button--size_large');
        expect(button[1]).toHaveClass('ButtonGroup__button--isBlock');
    });

    it('should support mixed element types', () => {
        const view = render(
            <ButtonGroup size="large" isBlock>
                <Button key="a">A button</Button>
                <Button key="b" href="#">
                    An anchor
                </Button>
                <Dropdown
                    key="c"
                    button={<Button>A dropdown button</Button>}
                    placement="bottom-end"
                    onChange={() => {}}
                >
                    <ListItem key="some-key">A list item</ListItem>
                </Dropdown>
            </ButtonGroup>
        );

        expect(view.container).toMatchSnapshot();
    });

    it('should render with conditional JSX', () => {
        const condition = false;
        const view = render(
            <ButtonGroup>
                <Button key="a">A button</Button>
                {condition ? <Button href="#">An anchor</Button> : null}
                {condition ? <Button key="b">A button</Button> : null}
            </ButtonGroup>
        );

        expect(view.container).toMatchSnapshot();
        expect(screen.queryByText('false')).not.toBeInTheDocument();
    });

    it('should pass main props, but not add styles if there is only 1 child', () => {
        const view = render(
            <ButtonGroup size="small" isBlock>
                <Button key="a">A button</Button>
            </ButtonGroup>
        );

        expect(view.container).toMatchSnapshot();
        const button = screen.getByRole('button');
        expect(button).toHaveClass('Button--isPrimary');
        expect(button).toHaveClass('Button--size_small');
        expect(button).toHaveClass('Button--isBlock');
    });

    it('should pass main props if child is array with size one', () => {
        const buttonContent = ['A button'];
        const view = render(
            <ButtonGroup isPrimary size="small" isBlock>
                {buttonContent.map((content) => (
                    <Button key="a">{content}</Button>
                ))}
            </ButtonGroup>
        );

        expect(view.container).toMatchSnapshot();
        const button = screen.getByRole('button');
        expect(button).toHaveClass('Button--isPrimary');
        expect(button).toHaveClass('Button--size_small');
        expect(button).toHaveClass('Button--isBlock');
    });
});
