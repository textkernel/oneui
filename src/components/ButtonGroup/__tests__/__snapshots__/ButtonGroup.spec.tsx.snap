// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ButtonGroup> that renders a button should add classes when props are changed 1`] = `
<div>
  <div
    class="ButtonGroup ButtonGroup--isBlock"
    role="group"
  >
    <button
      class="Button Button--isPrimary Button--size_large ButtonGroup__button ButtonGroup__button--isBlock ButtonGroup__button--isPrimary ButtonGroup__button--first"
      type="button"
    >
      A button
    </button>
    <button
      class="Button Button--isPrimary Button--size_large ButtonGroup__button ButtonGroup__button--isBlock ButtonGroup__button--isPrimary ButtonGroup__button--last"
      type="button"
    >
      Another button
    </button>
  </div>
</div>
`;

exports[`<ButtonGroup> that renders a button should pass main props if child is array with size one 1`] = `
<div>
  <div
    class="ButtonGroup ButtonGroup--isBlock"
    role="group"
  >
    <button
      class="Button Button--isPrimary Button--isBlock Button--size_small"
      type="button"
    >
      A button
    </button>
  </div>
</div>
`;

exports[`<ButtonGroup> that renders a button should pass main props, but not add styles if there is only 1 child 1`] = `
<div>
  <div
    class="ButtonGroup ButtonGroup--isBlock"
    role="group"
  >
    <button
      class="Button Button--isPrimary Button--isBlock Button--size_small"
      type="button"
    >
      A button
    </button>
  </div>
</div>
`;

exports[`<ButtonGroup> that renders a button should render default button correctly 1`] = `
<div>
  <div
    class="ButtonGroup"
    role="group"
  >
    <button
      class="Button Button--isPrimary ButtonGroup__button ButtonGroup__button--isPrimary ButtonGroup__button--first"
      type="button"
    >
      A button
    </button>
    <button
      class="Button Button--isPrimary ButtonGroup__button ButtonGroup__button--isPrimary ButtonGroup__button--last"
      type="button"
    >
      Another button
    </button>
  </div>
</div>
`;

exports[`<ButtonGroup> that renders a button should render with conditional JSX 1`] = `
<div>
  <div
    class="ButtonGroup"
    role="group"
  >
    <button
      class="Button Button--isPrimary ButtonGroup__button ButtonGroup__button--isPrimary ButtonGroup__button--first"
      type="button"
    >
      A button
    </button>
  </div>
</div>
`;

exports[`<ButtonGroup> that renders a button should support mixed element types 1`] = `
<div>
  <div
    class="ButtonGroup ButtonGroup--isBlock"
    role="group"
  >
    <button
      class="Button Button--isPrimary Button--size_large ButtonGroup__button ButtonGroup__button--isBlock ButtonGroup__button--isPrimary ButtonGroup__button--first"
      type="button"
    >
      A button
    </button>
    <a
      class="Button Button--isPrimary Button--size_large ButtonGroup__button ButtonGroup__button--isBlock ButtonGroup__button--isPrimary"
      href="#"
    >
      An anchor
    </a>
    <button
      aria-expanded="false"
      aria-haspopup="listbox"
      aria-labelledby="downshift-0-label downshift-0-toggle-button"
      class="Button Button--isPrimary Button--size_large ButtonGroup__button ButtonGroup__button--isBlock ButtonGroup__button--isPrimary ButtonGroup__button--last"
      id="downshift-0-toggle-button"
      type="button"
    >
      A dropdown button
    </button>
    <ul
      aria-labelledby="downshift-0-label"
      class="List"
      id="downshift-0-menu"
      role="listbox"
      tabindex="-1"
    />
  </div>
</div>
`;
